src/java.base/share/classes/sun/security/ssl/X509KeyManagerImpl.java

```diff
@@ -180,6 +299,107 @@ public String[] getServerAliases(String keyType, Principal[] issuers) {
     // implementation private methods
     //
 
+    private String removeAliasIndex (String alias){
+        if (alias == null) {
+            return null;
+        }
+
+        int firstDot = alias.indexOf('.');
+        int secondDot = alias.indexOf('.', firstDot + 1);
+        if ((firstDot == -1) || (secondDot == firstDot)) {
+            // invalid alias
+            return null;
+        }
+        return alias.substring(firstDot + 1);
+    }
+
+    private boolean areCertChainsEqual(String builderAlias,
+            X509Certificate[] cachedCerts) {
+        int aDot = builderAlias.indexOf('.');
+        int builderIndex = Integer.parseInt(builderAlias.substring(0, aDot));
+        String entryAlias = builderAlias.substring(aDot + 1);
+        try {
+            Builder builder = builders.get(builderIndex);
+            KeyStore keyStore = builder.getKeyStore();
+            if (keyStore == null) {
+                return false;
+            }
+            Certificate[] keyStoreCerts = keyStore.getCertificateChain(entryAlias);
+            if (keyStoreCerts == null || keyStoreCerts.length == 0) {
+                return false;
+            }
+            return Arrays.equals(cachedCerts, keyStoreCerts);
+        } catch (Exception e) {
+            if (SSLLogger.isOn && SSLLogger.isOn("keymanager")) {
+                SSLLogger.fine("KeyMgr: exception occurred during getKeyStore() " +
+                        "operation");
+            }
+            return false;
+        }
+    }
+
+    private boolean processCredentials(Builder builder, int builderIndex,
+            String alias) throws Exception {
+        KeyStore keyStore = builder.getKeyStore();
+        if (keyStore == null) {
+            return false;
+        }
+
+        Entry newEntry = keyStore.getEntry(alias,
+                builder.getProtectionParameter(alias));
+        if (!(newEntry instanceof PrivateKeyEntry privateKeyEntry)) {
+            return false;
+        }
+
+        PrivateKey privateKey = privateKeyEntry.getPrivateKey();
+        X509Certificate[] certs = (X509Certificate[]) privateKeyEntry.getCertificateChain();
+
+        X509Certificate cert = certs[0];
+        PublicKey publicKey = cert.getPublicKey();
+        if (!privateKey.getAlgorithm().equals(publicKey.getAlgorithm())) {
+            return false;
+        }
+
+        // Add to credentials map
+        String builderAlias = builderIndex + "." + alias;
+        X509Credentials cred = new X509Credentials(privateKey, certs);
+        credentialsMap.put(builderAlias, cred);
+
+        if (SSLLogger.isOn && SSLLogger.isOn("keymanager")) {
```

Comment:
The method `processCredentials` itself logs for the success, however the callers log for the failure (see line 169 and 388).
Especially, the logs on successful and failed cases display different contents.
The log for the failed case doesn't contain the builder index and entry alias.

Could it also log the failed cases in the body of this method?
That should log the messages before returning false.
Then, the callers may not need to log for the method.

Reply:
Done as suggested.